components/renku_data_services/base_models/core.py

```diff
@@ -3,41 +3,67 @@
 import re
 import unicodedata
 from dataclasses import dataclass, field
+from datetime import datetime
 from enum import Enum, StrEnum
-from typing import ClassVar, Optional, Protocol, Self
+from typing import ClassVar, Optional, Protocol, Self, TypeVar
 
 from sanic import Request
 
 from renku_data_services.errors import errors
 
 
-class Authenticator(Protocol):
-    """Interface for authenticating users."""
-
-    token_field: str
-
-    async def authenticate(self, access_token: str, request: Request) -> "APIUser":
-        """Validates the user credentials (i.e. we can say that the user is a valid Renku user)."""
-        ...
-
-
-@dataclass(kw_only=True)
+@dataclass(kw_only=True, frozen=True)
 class APIUser:
     """The model for a user of the API, used for authentication."""
 
+    id: str | None = None  # the sub claim in the access token - i.e. the Keycloak user ID
+    access_token: str | None = field(repr=False, default=None)
+    refresh_token: str | None = field(repr=False, default=None)
+    full_name: str | None = None
+    first_name: str | None = None
+    last_name: str | None = None
+    email: str | None = None
+    access_token_expires_at: datetime | None = None
     is_admin: bool = False
-    id: Optional[str] = None  # the sub claim in the access token - i.e. the Keycloak user ID
-    access_token: Optional[str] = field(repr=False, default=None)
-    full_name: Optional[str] = None
-    first_name: Optional[str] = None
-    last_name: Optional[str] = None
-    email: Optional[str] = None
 
     @property
     def is_authenticated(self) -> bool:
         """Indicates whether the user has successfully logged in."""
         return self.id is not None
 
+    def get_full_name(self) -> str | None:
+        """Generate the closest thing to a full name if the full name field is not set."""
+        full_name = self.full_name or " ".join(filter(None, (self.first_name, self.last_name)))
+        if len(full_name) == 0:
+            return None
+        return full_name
+
+
+@dataclass(kw_only=True, frozen=True)
+class AuthenticatedAPIUser(APIUser):
+    """The model for a an authenticated user of the API."""
+
+    id: str
+    email: str
+    access_token: str = field(repr=False)
+    refresh_token: str | None = field(default=None, repr=False)
+    full_name: str | None = None
+    first_name: str | None = None
+    last_name: str | None = None
+
+
+@dataclass(kw_only=True, frozen=True)
+class AnonymousAPIUser(APIUser):
+    """The model for an anonymous user of the API."""
+
+    id: str
+    is_admin: bool = field(init=False, default=False)
```

Comment:
Shouldn't repr also be set to false ?

```suggestion
    is_admin: bool = field(init=False, default=False, repr=False)
```

Reply:
I dont see why we need to hide this? It is nice to see whether the user is or isnt an admin. There is nothing secret here.

Reply:
My thinking is that as an anonymous users, only the bare minimum of information should be returned.

Reply:
But this is only what will show up in the logs. When we use the class we can still access that. I would just leave this as is. The most important thing is that with `init=False` and the `frozen=True` it is very hard for someone in the code to make an anonymous user where is_admin is True. I would rather be able to see in the logs that this anonymous user really had the field set to False in the logs.

Reply:
Fair enough